### ISSUE TITLE
Blurred image only shown briefly

### ISSUE THREAD

**Issue by JoshuaWalsh** [gatsbyjs/gatsby#14325]
## Description

I was previously experiencing #14186. After upgrading to gatsby-remark-images 3.x the situation has improved, but is still not working correctly.

The blurred image only appears for a very short amount of time, and then is covered with a blank white rectangle.

### Expected result

When I load my page on a slow network connection, I expect the placeholder image to be displayed until the full image has loaded.

### Actual result

![blurUp](https://user-images.githubusercontent.com/2290576/58370621-033bbf00-7f4c-11e9-97bd-728eb8e2923c.gif)

(This GIF is a real-time recording using Firefox's GPRS network throttling preset)

### Environment

```

  System:
    OS: Windows 10
    CPU: (16) x64 AMD Ryzen 7 1800X Eight-Core Processor         
  Binaries:
    Yarn: 1.13.0 - C:\Program Files (x86)\Yarn\bin\yarn.CMD
    npm: 6.5.0-next.0 - C:\Program Files\nodejs\npm.CMD
  Languages:
    Python: 2.7.15
  Browsers:
    Edge: 44.17763.1.0
  npmPackages:
    gatsby: ^2.2.6 => 2.2.6 
    gatsby-image: ^2.0.22 => 2.0.25 
    gatsby-plugin-canonical-urls: ^2.0.8 => 2.0.8 
    gatsby-plugin-catch-links: ^2.0.9 => 2.0.9 
    gatsby-plugin-excerpts: ^1.0.2 => 1.0.2 
    gatsby-plugin-favicon: ^3.1.5 => 3.1.5 
    gatsby-plugin-feed: ^2.0.8 => 2.0.11 
    gatsby-plugin-google-analytics: ^2.0.5 => 2.0.8 
    gatsby-plugin-manifest: ^2.0.5 => 2.0.12 
    gatsby-plugin-meta-redirect: ^1.1.1 => 1.1.1 
    gatsby-plugin-offline: ^2.0.5 => 2.0.20 
    gatsby-plugin-prefetch-google-fonts: ^1.1.1 => 1.1.1 
    gatsby-plugin-react-helmet: ^3.0.0 => 3.0.5 
    gatsby-plugin-remove-serviceworker: ^1.0.0 => 1.0.0 
    gatsby-plugin-root-import: ^2.0.4 => 2.0.4 
    gatsby-plugin-s3: ^0.2.5 => 0.2.5 
    gatsby-plugin-sass: ^2.0.7 => 2.0.7 
    gatsby-plugin-sharp: ^2.0.30 => 2.0.30 
    gatsby-plugin-sitemap: ^2.0.4 => 2.0.4 
    gatsby-plugin-typography: ^2.2.0 => 2.2.3 
    gatsby-redirect-from: ^0.1.1 => 0.1.1 
    gatsby-remark-autolink-headers: ^2.0.12 => 2.0.12 
    gatsby-remark-behead: 0.0.3 => 0.0.3 
    gatsby-remark-copy-linked-files: ^2.0.5 => 2.0.8 
    gatsby-remark-custom-blocks: ^2.0.3 => 2.0.3 
    gatsby-remark-embed-youtube: 0.0.7 => 0.0.7 
    gatsby-remark-external-links: 0.0.4 => 0.0.4 
    gatsby-remark-images: ^3.0.0 => 3.0.12 
    gatsby-remark-katex: ^3.0.1 => 3.0.1 
    gatsby-remark-prismjs: ^3.0.0 => 3.1.4 
    gatsby-remark-responsive-iframe: ^2.0.5 => 2.0.8 
    gatsby-remark-smartypants: ^2.0.5 => 2.0.7 
    gatsby-source-filesystem: ^2.0.27 => 2.0.27 
    gatsby-transformer-remark: ^2.1.6 => 2.1.17 
    gatsby-transformer-sharp: ^2.1.3 => 2.1.9 

```

**Comment by KyleAMathews** [comment]
What page is that? Does it only happen on Firefox?

**Comment by JoshuaWalsh** [comment]
In Firefox the preview appears briefly. In Chrome the blurred image doesn't appear at all, the white rectangle is instant. (Using a custom throttling profile of 50kb/s down, 20kb/s up, 500ms latency)

The page is [here](https://blog.joshwalsh.me/aws-gatsby/).

**Comment by KyleAMathews** [comment]
So this only happens on throttled connections? I just opened the page and
the image loaded as expected.

**Comment by KyleAMathews** [comment]
Oh and does the image eventually load?

**Comment by JoshuaWalsh** [comment]
Yes, the image eventually loads. It happens on all connections, but throttling makes it easier to see. When you say the image loaded as expected, do you mean you saw the blurred/traced version of the image while it was loading?

**Comment by freiksenet** [comment]
I have it happening on Chrome too, with throttled and not-throttled connection. @JoshuaWalsh I wonder if you could give us access to the website source? Alternatively I wonder if you could extract a small reproduction case. Thank you very much!

**Comment by JoshuaWalsh** [comment]
I have managed to reproduce it in an extremely minimal site. Usually when I cut things down to a minimal example I learn what was causing them along the way, but not this time. There's not much left of the site but it still doesn't work.

Steps to reproduce:

1. Download [the example](randomshit.joshwalsh.me/gatsby-remark-images-reprex.zip) and unzip it somewhere
2. In that directory, run `npm ci`. This will ensure you install the same version of dependencies as I have.
3. Run `gatsby build`.
4. Host the public folder somewhere. I'm using an Apache vhost on my local machine, but nginx would also work.
5. Navigate to http://placeyouarehostingit.local/test/ in Chrome or Firefox
6. Apply throttling with devtools to ensure that the image doesn't just load instantly. (See throttling presets from my original post)
7. Reload the page. Observe that the traced image appears initially, but disappears as soon as the alt text appears. Meanwhile the image takes a bit longer to load.

**Comment by KyleAMathews** [comment]
Sorry for not catching this earlier. I wasn't paying attention and thought you were talking about gatsby-image.

This is actually the _currently_ expected behavior for gatsby-remark-images. It's not as sophisticated as gatsby-image so doesn't do lazy loading & fade-in of the image when it loads. It just sets the placeholder and as soon as the real image starts loading, that gets rendered above the placeholder causing the behavior you're seeing on slower connections. I'd love someone to take on porting the functionality from gatsby-image to gatsby-remark-images. The main difficulty is that gatsby-remark-images is just HTML not a React component. The main way around this I think is to have the plugin add some JS to pages that takes over the HTML (old skoll jquery style) and does the lazy loading.

Interested in trying this? We could create a new issue to discussion implementation ideas.

**Comment by JoshuaWalsh** [comment]
I think this is important to implement. The current blur-up implementation is not particularly useful given that it will be displayed for less than a second on a dial-up connection, and probably never on even a 2G mobile connection.

I'm interested in trying this. I have two ideas:

Option #1: Use rehype-react in the same way that `gatsby-remark-component` does.

Pros:

 - Then we can just use gatsby-image and avoid reinventing the wheel

Cons:

 - Requires people to change their page code from using dangerouslySetInnerHTML to using htmlAst and rehype-react renderAst. (Maybe we can provide a graceful fallback so the current behaviour is preserved if the user continues using dangerouslySetInnerHTML?)

Option #2: Provide some JavaScript which runs on the client and manages markdown image loading

Pros:

 - Works with users' existing code

Cons:

 - If I'd tried to do something like this in AngularJS I would've been told it's "not the Angular way". I'm assuming the same probably applies to React. The JS code would essentially sit outside of the React ecosystem.

I think we would need to do some tests, but I think IF we can provide a graceful fallback for users with old code then Option #1 is preferable.

**Comment by KyleAMathews** [comment]
Option 2 is totally acceptable.

**Comment by JoshuaWalsh** [comment]
Actually, I've done some further digging and the explanation for this behaviour is slightly different.

I noticed that if instead of using Firefox's network throttling, I delayed ONLY the image request (I used mod_ratelimit but you could also use Fiddler to achieve this) then the traced version did appear and then faded to the full image once it was loaded. I was pretty surprised by this, but sure enough `gatsby-browser.js` [has code to handle it](https://github.com/gatsbyjs/gatsby/blob/72077527b4acd3f2109ed5a2fcb780cddefee35a/packages/gatsby-remark-images/src/gatsby-browser.js#L20).

So then the question is, why doesn't this work if the whole connection is slow? Why does it only work if the image loads slowly? The answer can be found by looking at the Waterfall:

![image](https://user-images.githubusercontent.com/2290576/58630879-196ac600-8324-11e9-9043-063f01916362.png)

The issue is that the image finishes loading before gatsby-remark-images' clientside code.

Maybe this is an edge case because my images are simple and so end up being compressed very small, but given that the images start loading halfway through the page load while the JavaScript only starts loading after all the document is fully loaded, and also given that app-*.js weighs in at a hefty 185KiB for my basic website, I think it's fairly easy to have images small enough that they finish loading before the JavaScript does.

I'm sure adding lazy-loading to the images would fix the issue, but only because it would make the images load slower.

**Comment by mohoff** [comment]
I'm having a very similar issue (see https://github.com/gatsbyjs/gatsby/issues/14486) and i'm not using `gatsby-remark-images`. The blurred images shortly pop up during page load and disappear then completely.

@JoshuaWalsh Does it work on `gatsby develop` as well for you?

**Comment by JoshuaWalsh** [comment]
Its hard for me to test with `gatsby develop` due to commons.js taking an incredibly long time to load with throttling settings enabled. It seems the fade-up does work with `gatsby develop`. This is expected given that the JS loads before the content in this mode.

If you aren't using `gatsby-remark-images`, are you just using `gatsby-images`? Interesting if the bug also affects that plugin.

**Comment by JoshuaWalsh** [comment]
I found a solution to this issue, but it's not perfect. I'd like some feedback before I make a PR.

Removing the box-shadow inline style from the img element fixes the issue. But the alt text appears over the placeholder image. It's nearly impossible to read due to it being on top of the image, and it makes the image look ugly.

So I also applied `color: transparent` to the img element. This makes the alt text invisible, but it will still be read by screen readers. But it also means if the image fails to load the alt text will be invisible.

So I'm considering also adding some JS to check for images that failed to load and on them set the background to white and the color to black, so the alt text will be readable.

What about people with JS disabled? I also copied the alt text into the title, which means it's possible (but inconvenient) for people with JS disabled to see it too. Plus people who can see the images can also see the text.

Finally, I left all the blur-up code in place so for people that are able to load the JS before the images the animation still works.

Overall I feel like I've made good compromises between percieved-performance, accessibility, aesthetics, and functionality.

@KyleAMatthews Should I submit a PR with these changes? Or is there some kind of conceptual flaw I've missed?
